source/common/protobuf/utility.cc

```diff
@@ -506,6 +508,114 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+std::unique_ptr<Protobuf::Message> typeUrlToNewMessage(Protobuf::MessageFactory& message_factory,
+                                                       absl::string_view type_url) {
+  const absl::string_view type_name = TypeUtil::typeUrlToDescriptorFullName(type_url);
+  const Protobuf::Descriptor* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(
+          static_cast<std::string>(type_name));
+  return std::unique_ptr<Protobuf::Message>(
+      descriptor == nullptr ? nullptr : message_factory.GetPrototype(descriptor)->New());
+}
+
+// Recursive helper method for MessageUtil::redact() below. Note that we have to keep track of
+// whether an ancestor was marked as `sensitive`, not just the field, because of cases like
+// `TlsContext::private_key`, which is of type `core.DataSource` rather than `string`.
```

Comment:
Hmm do we want to redact if the DataSource is pointing to a file rather than inline string/bytes?

Reply:
This approach doesn't give us any way to differentiate, unfortunately.